cmds/boot/boot.go

```diff
@@ -133,107 +100,171 @@ func getSupportedFilesystem() ([]string, error) {
+	if err != nil {
+		return nil, "", "", err
 	}
-	returnValue = false
-	return returnValue
+
+	// it's easier we think to do include processing here.
+	lines := strings.Split(string(isolinux), "\n")
+	var result []string
+	for _, l := range lines {
+		f := strings.Fields(l)
+		if len(f) == 0 {
+			continue
+		}
+		switch f[0] {
+		default:
+			result = append(result, l)
+		case "include":
+			i, _, _, err := loadISOLinux(dir, f[1])
+			if err != nil {
+				return nil, "", "", err
+			}
+			result = append(result, i...)
+			// If the string contain include
+			// we must call back as to read the content of the file
+		}
+	}
+	return result, dir, dir, nil
 }
 
 // checkBootEntry is looking for grub.cfg file
 // and return absolute path to it
-func checkBootEntry(mountPoint string) ([]byte, string) {
-	grub, err := ioutil.ReadFile(filepath.Join(mountPoint, "/boot/grub/grub.cfg"))
+func checkBootEntry(mountPoint string) ([]string, string, string, error) {
+	grub, err := ioutil.ReadFile(filepath.Join(mountPoint, "boot/grub/grub.cfg"))
 	if err == nil {
-		return grub, filepath.Join(mountPoint, "/boot/grub")
+		return strings.Split(string(grub), "\n"), filepath.Join(mountPoint, "/isolinux"), mountPoint, nil
 	}
-	return grub, ""
-
+	return loadISOLinux(filepath.Join(mountPoint, "isolinux"), "isolinux.cfg")
 }
 
 // getFileMenuContent is parsing a grub.cfg file
-// input: absolute directory path to grub.cfg
-// output: Return a list of strings with the following format
-//	 line[3*x] - menuconfig
-//	 line[3*x+1] - linux kernel + boot options
-// 	 line[3*x+2] - initrd
-// and the default boot entry configured into grub.cfg
-func getFileMenuContent(file []byte) ([]string, int, error) {
-	var returnValue []string
-	var err error
-	var status int
-	var intReturn int
-	intReturn = 0
-	status = 0
-	// When status = 0 we are looking for a menu entry
-	// When status = 1 we are looking for a linux entry
-	// When status = 2 we are looking for a initrd entry
-	var trimmedLine string
-	s := string(file)
-	for _, line := range strings.Split(s, "\n") {
-		trimmedLine = strings.TrimSpace(line)
-		trimmedLine = strings.Join(strings.Fields(trimmedLine), " ")
-		if strings.HasPrefix(trimmedLine, "#") {
-			continue
-		}
-		if (strings.HasPrefix(trimmedLine, "set default=")) && (status == 0) {
-			fmt.Sscanf(trimmedLine, "set default=\"%d\"", &intReturn)
+// output: bootEntries
+// grub parsing is a good deal messier than it looks.
+// This is a simple parser will fail on anything tricky.
+func getBootEntries(lines []string) (map[string]*bootEntry, map[string]string, error) {
+	// There are two ways to reference a bootEntry: its order in the file and its
+	// name.
+	var (
+		lineno   int
+		line     string
+		err      error
+		curEntry string
+		numEntry int
+		b        = make(map[string]*bootEntry)
+		v        = make(map[string]string)
+		f        []string
+		be       *bootEntry
+	)
+	defer func() {
+		switch err := recover().(type) {
```

Comment:
:( what panics? I'd prefer fixing the panic

Reply:
The parser. This is recovery for a parser, which makes the parser simpler; I grabbed this technique from a talk rob gave a few years back and I've used it in the shell and the ip command.